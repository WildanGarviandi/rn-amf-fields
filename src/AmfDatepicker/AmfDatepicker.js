import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { DatePickerAndroid, StyleSheet, Text, TouchableOpacity, View } from 'react-native'
import moment from 'moment'

class AmfDatepicker extends Component {
  constructor(props) {
    super(props)

    this.state = {
      valid: false,
      errorMessage: ''
    }
  }

  async openDatepicker(currentDate) {
    try {
      let options = {
        date : moment.unix(currentDate).valueOf()
      }
      const {action, year, month, day} = await DatePickerAndroid.open(options)

      if (action !== DatePickerAndroid.dismissedAction) {
        let formattedDate = moment({y: year, M: month, d: day}).format('YYYY-MM-DD')
        this.props.onSelect(formattedDate)
      } else {
        this.props.onSelect('')
      }

    } catch ({code, message}) {
      console.log(message)
    }
  }

  componentDidMount() {
    this.validate()
  }

  componentDidUpdate(prevProps) {
    if (this.props.selectedDate !== prevProps.selectedDate) {
      this.validate()
    }
  }

  onPress = () => {
    this.openDatepicker(moment(this.props.selectedDate, 'YYYY-MM-DD').unix())
  }

  inputValid = () => {
    const { validation, label } = this.props
    if (!validation) return ({status: true})

    // check if pristine
    if (this.props.selectedDate === null || this.props.selectedDate === undefined) return ({status: true})

    // check whether input is required
    if (validation.required) {
      if (this.props.selectedDate === '') return ({status: false, message: label + ' tidak boleh kosong'})
    }

    return ({status: true})
  }

  validate = () => {
    let validationObject = this.inputValid()
    this.setState({valid: validationObject.status, errorMessage: validationObject.message})
  }

  render() {
    let dateFormatted
    if (this.props.selectedDate && this.props.selectedDate !== '') dateFormatted = moment(this.props.selectedDate, 'YYYY-MM-DD').format('DD/MMM/YYYY')
    else dateFormatted = 'Pilih Tgl'

    let validationStyle = {}
    let validationText
    if (!this.state.valid) {
      validationStyle = {borderColor: 'red', borderWidth: 1}
      validationText = <Text style={styles.errorText}>{this.state.errorMessage}</Text>
    }

    return (
      <View>
        <View style={ this.props.layout === 'horizontal' ? styles.container : null}>
          <Text style={styles.label}>{this.props.label}</Text>
          <TouchableOpacity style={[styles.calendarButton, validationStyle]}
            onPress={this.onPress}
          >
            <Text>{ dateFormatted }</Text>
          </TouchableOpacity>
        </View>
        { validationText }
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container : {
    flexDirection: 'row',
    alignItems: 'center'
  },
  calendarButton: {
    backgroundColor: 'white',
    elevation: 4,
    borderRadius: 16,
    width: 100,
    padding: 7,
    alignItems: 'center'
  },
  errorText: {
    color: 'red',
    margin: 4,
    marginBottom: 0
  },
  label: {
    color: '#454545',
    marginRight: 16
  },
})

AmfDatepicker.propTypes = {
  label: PropTypes.string,
  layout: PropTypes.oneOf(['vertical', 'horizontal']),
  onSelect : PropTypes.func.isRequired,
  selectedDate : PropTypes.string,
  validation: PropTypes.object
}

export default AmfDatepicker
