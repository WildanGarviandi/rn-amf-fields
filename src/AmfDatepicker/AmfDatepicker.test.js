import React from 'react'
import AmfDatepicker from './AmfDatepicker'
import renderer from 'react-test-renderer'

// jest.mock('./DatepickerAndroid')

const onSelect = jest.fn()

test('AmfDatepicker without validation', () => {
  let props = { onSelect }

  const component = renderer.create(<AmfDatepicker {...props} />)
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('AmfDatepicker with unselected date', () => {
  let props = { onSelect, selectedDate: null }
  let propsEmpty = { ...props, selectedDate: '' }
  
  const component = renderer.create(<AmfDatepicker {...props} />)
  const componentEmpty = renderer.create(<AmfDatepicker {...propsEmpty} />)
  let tree = component.toJSON()
  let treeEmpty = componentEmpty.toJSON()
  expect(tree).toMatchSnapshot()
  expect(treeEmpty).toMatchSnapshot()
})

test('AmfDatepicker with layout', () => {
  let props = {
    onSelect,
    layout: 'vertical'
  }
  let propsHorizontal = {
    ...props, layout: 'horizontal'
  }

  const component = renderer.create(<AmfDatepicker {...props} />)
  const componentHorizontal = renderer.create(<AmfDatepicker {...propsHorizontal} />)

  let tree = component.toJSON()
  let treeHorizontal = component.toJSON()
  expect(tree).toMatchSnapshot()
  expect(treeHorizontal).toMatchSnapshot()
})

test('test required validation of AmfDatepicker', () => {
  let props = { onSelect, validation: { required: true } }
  let propsNull = { ...props, selectedDate: null }
  let propsEmpty = { ...props, selectedDate: '' }
  let propsNotEmpty = { ...props, selectedDate: '2017-07-07' }

  const component = renderer.create(<AmfDatepicker {...props} />)
  const componentNull = renderer.create(<AmfDatepicker {...propsNull} />)
  const componentEmpty = renderer.create(<AmfDatepicker {...propsEmpty} />)
  const componentNotEmpty = renderer.create(<AmfDatepicker {...propsNotEmpty} />)

  let tree = component.toJSON()
  let treeNull = componentNull.toJSON()
  let treeEmpty = componentEmpty.toJSON()
  let treeNotEmpty = componentNotEmpty.toJSON()

  expect(tree).toMatchSnapshot()
  expect(treeNull).toMatchSnapshot()
  expect(treeEmpty).toMatchSnapshot()
  expect(treeNotEmpty).toMatchSnapshot()
})

test('AmfDatepicker test onPress', () => {
  let props = { onSelect }
  const component = new AmfDatepicker(props)
  component.openDatepicker = jest.fn()
  component.onPress()
  expect(component.openDatepicker).toHaveBeenCalled()
})

test('AmfDatepicker test componentDidUpdate', () => {
  let props = { onSelect }
  const component = new AmfDatepicker(props)
  component.validate = jest.fn()
  component.props.selectedDate = '2017-07-07'
  component.componentDidUpdate({selectedDate: '2017-07-06'})
  expect(component.validate).toHaveBeenCalled()
})

// test('AmfDatepicker test openDatepicker function', async () => {
//   let props = { onSelect }

//   const component = new AmfDatepicker(props)
//   component.openDatepicker('2017-07-07')
//   await expect(props.onSelect).resolves.toHaveBeenCalled()
// })
