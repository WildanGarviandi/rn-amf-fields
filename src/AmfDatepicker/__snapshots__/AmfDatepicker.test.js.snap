// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmfDatepicker with layout 1`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#454545",
          "marginRight": 16,
        }
      }
    />
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 0,
          "elevation": 4,
          "opacity": 1,
          "padding": 7,
          "width": 120,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
      >
        Pilih Tgl
      </Text>
    </View>
  </View>
</View>
`;

exports[`AmfDatepicker with layout 2`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#454545",
          "marginRight": 16,
        }
      }
    />
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 0,
          "elevation": 4,
          "opacity": 1,
          "padding": 7,
          "width": 120,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
      >
        Pilih Tgl
      </Text>
    </View>
  </View>
</View>
`;

exports[`AmfDatepicker with unselected date 1`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#454545",
          "marginRight": 16,
        }
      }
    />
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 0,
          "elevation": 4,
          "opacity": 1,
          "padding": 7,
          "width": 120,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
      >
        Pilih Tgl
      </Text>
    </View>
  </View>
</View>
`;

exports[`AmfDatepicker with unselected date 2`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#454545",
          "marginRight": 16,
        }
      }
    />
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 0,
          "elevation": 4,
          "opacity": 1,
          "padding": 7,
          "width": 120,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
      >
        Pilih Tgl
      </Text>
    </View>
  </View>
</View>
`;

exports[`AmfDatepicker without validation 1`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#454545",
          "marginRight": 16,
        }
      }
    />
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 0,
          "elevation": 4,
          "opacity": 1,
          "padding": 7,
          "width": 120,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
      >
        Pilih Tgl
      </Text>
    </View>
  </View>
</View>
`;

exports[`test required validation of AmfDatepicker 1`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#454545",
          "marginRight": 16,
        }
      }
    />
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 0,
          "elevation": 4,
          "opacity": 1,
          "padding": 7,
          "width": 120,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
      >
        Pilih Tgl
      </Text>
    </View>
  </View>
</View>
`;

exports[`test required validation of AmfDatepicker 2`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#454545",
          "marginRight": 16,
        }
      }
    />
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 0,
          "elevation": 4,
          "opacity": 1,
          "padding": 7,
          "width": 120,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
      >
        Pilih Tgl
      </Text>
    </View>
  </View>
</View>
`;

exports[`test required validation of AmfDatepicker 3`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#454545",
          "marginRight": 16,
        }
      }
    />
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderColor": "red",
          "borderRadius": 0,
          "borderWidth": 1,
          "elevation": 4,
          "opacity": 1,
          "padding": 7,
          "width": 120,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
      >
        Pilih Tgl
      </Text>
    </View>
  </View>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "red",
        "margin": 4,
        "marginBottom": 0,
      }
    }
  >
    undefined tidak boleh kosong
  </Text>
</View>
`;

exports[`test required validation of AmfDatepicker 4`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#454545",
          "marginRight": 16,
        }
      }
    />
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 0,
          "elevation": 4,
          "opacity": 1,
          "padding": 7,
          "width": 120,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
      >
        Pilih Tgl
      </Text>
    </View>
  </View>
</View>
`;
